source/src_lcao/LCAO_gen_fixedH.cpp

```diff
@@ -1135,6 +1143,214 @@ void LCAO_gen_fixedH::build_Nonlocal_beta_new() //update by liuyu 2021-04-07
+                                std::vector<double> nlm2 = nlm_tot_thread[ad2_count][iw2_all];
+                            #else 
+                                std::vector<double> nlm1 = nlm_tot[ad1_count][iw1_all];
+                                std::vector<double> nlm2 = nlm_tot[ad2_count][iw2_all];
+                            #endif 
+
+                            assert(nlm1.size()==nlm2.size());
+                            #ifdef _OPENMP
+                                double nlm_thread=0.0;
+                            #else 
+                                double nlm=0.0;
+                            #endif
+                            const int nproj = GlobalC::ucell.infoNL.nproj[T0];
+                            int ib = 0;
+                            for(int nb = 0; nb < nproj; nb++)
+                            {
+                                const int L0 = GlobalC::ucell.infoNL.Beta[T0].Proj[nb].getL();
+                                for(int m=0;m<2*L0+1;m++)
+                                {
+                                    #ifdef _OPENMP
+                                        nlm_thread += nlm1[ib]*nlm2[ib]*GlobalC::ucell.atoms[T0].dion(nb,nb);
+                                    #else 
+                                        nlm += nlm1[ib]*nlm2[ib]*GlobalC::ucell.atoms[T0].dion(nb,nb);
+                                    #endif
+                                    ib+=1;
+                                }
+                            }
+                            assert(ib==nlm1.size());
+
+                            const int ir = GlobalC::ParaO.trace_loc_row[ iw1_all ];
+                            const int ic = GlobalC::ParaO.trace_loc_col[ iw2_all ];
+                            long index=0;
+                            if(GlobalV::KS_SOLVER=="genelpa" || GlobalV::KS_SOLVER=="scalapack_gvx")
+                            {
+                                index=ic*GlobalC::ParaO.nrow+ir;
+                            }
+                            else
+                            {
+                                index=ir*GlobalC::ParaO.ncol+ic;
+                            }
+                            #ifdef _OPENMP
+                                Nonlocal_thread[index] += nlm_thread;
+                            #else 
+                                GlobalC::LM.set_HSgamma(iw1_all,iw2_all,nlm,'N');
+                            #endif
+                        }//iw2
+                    }//iw1
+                }//ad2
+            }//ad1
+        }//end iat
+
+        #ifdef _OPENMP
+            #pragma omp critical(cal_nonlocal)
+            {
+                for(int i=0; i<GlobalC::ParaO.nloc; i++)
+                {
+                    GlobalC::LM.Hloc_fixed[i] += Nonlocal_thread[i];
+                }
+            }
+            delete[] Nonlocal_thread;
+        #endif
+#ifdef _OPENMP
+    }
+#endif
+
+#ifdef __MKL
+    mkl_set_num_threads(mkl_threads);
```

Comment:
This function takes precedence over the MKL_NUM_THREADS environment variable.
It may cause thread oversubscribing when conbined with multiple MPI processes within 1 node. Also see line 1147.